autoload/maktaba/syscall.vim

```diff
@@ -89,6 +103,59 @@ function! maktaba#syscall#DoCall() abort dict
   return l:return_data
 endfunction
 
+
+""
+" @public
+" Returns whether the current vim session supports asynchronous calls.
+function! maktaba#syscall#IsAsyncAvailable()
+  return !s:async_disabled && !empty(v:servername) && has('clientserver')
+endfunction
+
+
+""
+" @private
+" @dict Syscall
+" Calls |system()| asynchronously, and invokes a @function(this.callback) once
+" the command completes, passing in stdout, stderr and exit code to it.
+" The specific implementation for @function(#CallAsync).
+function! maktaba#syscall#DoCallAsync() abort dict
+  if !maktaba#syscall#IsAsyncAvailable()
+    if self.allow_sync_fallback
+      let l:return_data = self.Call()
+      let l:return_data.status = 0
+      call maktaba#function#Call(self.callback, [s:CurrentEnv(), l:return_data])
+      return {}
+    else
+      if empty(v:servername)
+        throw maktaba#error#Message('ShellError', 'Cannot run async commands, '
+            \ 'no --servername flag passed to vim. See :help servername.')
+      elseif !has('clientserver')
+        throw maktaba#error#Message('ShellError', 'Cannot run async commands,' .
+            \ ' vim was compiled without +clientserver. See :help clientserver')
+      endif
+    endif
+  endif
+  let l:error_file = tempname()
+  let l:output_file = tempname()
+  let l:callback_cmd = join([
+        \ v:progname,
+        \ '--servername ' . v:servername,
+        \ '--remote-expr',
+        \ '"maktaba#syscall#AsyncDone(',
```

Comment:
`printf()` would make this easier to read since there's non-trivial structure mixed in with all this concatenation:
```vim
printf('"maktaba#syscall#AsyncDone(''%s'', ''%s'', $?)"',
    \ l:output_file, l:error_file)
```